src/app/Library/CrudPanel/Traits/Update.php

```diff
@@ -94,103 +90,174 @@ private function getModelAttributeValue($model, $field)
+        $relation = $relatedModel->{$relationMethod}();
+        $relationType = Str::afterLast(get_class($relation), '\\');
 
-        switch ($relation_type) {
+        switch ($relationType) {
             case 'MorphMany':
             case 'HasMany':
             case 'BelongsToMany':
             case 'MorphToMany':
-                // use subfields aka. pivotFields
-                if (! isset($field['subfields'])) {
-                    return $related_model->{$relation_method};
-                }
-                // we want to exclude the "self pivot field" since we already have it.
-                $pivot_fields = Arr::where($field['subfields'], function ($item) use ($field) {
-                    return $field['name'] != $item['name'];
-                });
-
-                $related_models = $related_model->{$relation_method};
-                $result = [];
-
-                // for any given model, we grab the attributes that belong to our pivot table.
-                foreach ($related_models as $related_model) {
-                    $item = [];
-                    switch ($relation_type) {
+                $relationModels = $relatedModel->{$relationMethod};
+                $result = collect();
+
+                foreach ($relationModels as $model) {
+                    // when subfields are NOT set we don't need to get any more values
+                    // we just return the plain models as we only need the ids
+                    if (! isset($field['subfields'])) {
+                        $result->push($model);
+                        continue;
+                    }
+                    // when subfields are set we need to parse their values so they can be displayed
+                    switch ($relationType) {
                         case 'HasMany':
                         case 'MorphMany':
-                            $result[] = $related_model->getAttributes();
+                            // we will get model direct attributes and merge with subfields values.
+                            $directAttributes = $this->getModelWithFakes($model)->getAttributes();
+                            $result->push(array_merge($directAttributes, $this->getSubfieldsValues($field['subfields'], $model)));
                             break;
 
                         case 'BelongsToMany':
                         case 'MorphToMany':
-                            $item = $related_model->pivot->getAttributes();
-                            $item[$field['name']] = $related_model->getKey();
+                            // for any given model, we grab the attributes that belong to our pivot table.
+                            $item = $model->pivot->getAttributes();
+                            $item[$relationMethod] = $model->getKey();
                             $result[] = $item;
                             break;
                     }
                 }
 
                 return $result;
```

Comment:
The fact that we have to convert the collection back to an array ALL multiple field types... it triggers my spidey-sense. Let's see if that's really needed. 

What I understand is that:
- previously, all those fields expected `$field['value']` to be an array;
- now, starting with the 4.2 branch, but on this branch too, `$field['value']` can also be a collection;

**Why? Why did we introduce collection-support on fields, on the 4.2 branch? Is it really _needed_? Are there benefits to it?** If there are no real benefits to it, I'd rather we convert the collection to array upstream (one time), instead of doing in in 5 places in the blade files. 

I believe we can do that here, with something like this:

```suggestion
                // convert the collection back to an array
                if (!empty($result)) {
                    if (is_a($result, \Illuminate\Support\Collection::class)) {
                        $result = $result->pluck($field['attribute'])->toArray();
                    }
                }
                
                return $result;
```

And then remove the support for collections from the fields themselves.

**Am I wrong? Can we do that @pxpm ?** Let me know if you want me to jump on a call, if it's easier to explain that way.

Reply:
I'm asking because... fields are supposed to be something _simple to create_. 

Custom multiple inputs shouldn't have to treat a bunch of cases themselves (is array, is collection, is db collection, is string), they should know "_ok, value is this format, I parse it, and make sure the input is in the right format_".

Maybe we've offset this complexity to the fields, when we should have treated it in the Update trait.

Reply:
Solved